Error while checking build/O2/fullCI_slc9 for b247f5cbc1b5f59aaf0a5a73a7a69a8eab5b0292 at 2025-09-28 13:16:
```
## sw/BUILD/O2-full-system-test-latest/log
task timeout reached .. killing all processes
[38881:BadMapCalibSpec]: [12:36:57][ERROR] Insufficient statistics: 0 entries in lowE histo, do nothing


## sw/BUILD/o2checkcode-latest/log
--
========== List of errors found ==========
++ GRERR=0
++ grep -v clang-diagnostic-error error-log.txt
++ grep ' error:'
++ GRERR=1
++ [[ 1 == 0 ]]
++ mkdir -p /sw/INSTALLROOT/3068eca543fac89842878422b043d539093ad03a/slc9_x86-64/o2checkcode/1.0-local363/etc/modulefiles
++ cat
--
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/14427/b247f5cbc1b5f59aaf0a5a73a7a69a8eab5b0292/build_O2_fullCI_slc9/pretty.html).

--- a/DataFormats/Detectors/TPC/CMakeLists.txt
+++ b/DataFormats/Detectors/TPC/CMakeLists.txt
@@ -40,6 +40,7 @@ o2_add_library(
 
 o2_target_root_dictionary(
   DataFormatsTPC
+  EXTRA_PATCH src/VectorPadflagsCustomStreamer.cxx
   HEADERS include/DataFormatsTPC/ClusterGroupAttribute.h
           include/DataFormatsTPC/ClusterNative.h
           include/DataFormatsTPC/ClusterNativeHelper.h

--- a/DataFormats/Detectors/TPC/src/DataFormatsTPCLinkDef.h
+++ b/DataFormats/Detectors/TPC/src/DataFormatsTPCLinkDef.h
@@ -22,6 +22,7 @@
 #pragma link C++ class o2::tpc::ClusterHardwareContainerFixedSize < 8192> + ;
 #pragma link C++ class o2::tpc::ClusterNativeContainer + ;
 #pragma link C++ class o2::tpc::Digit + ;
+// pragma link C++ enum o2::tpc::PadFlags +;                // enum itself
 #pragma link C++ class o2::tpc::ZeroSuppressedContainer8kb + ;
 #pragma link C++ class std::vector < o2::tpc::ClusterNative> + ;
 #pragma link C++ class std::vector < o2::tpc::ClusterNativeContainer> + ;
@@ -29,6 +30,7 @@
 #pragma link C++ class std::vector < o2::tpc::ClusterHardwareContainerFixedSize < 8192>> + ;
 #pragma link C++ class std::vector < o2::tpc::ClusterHardwareContainer8kb> + ;
 #pragma link C++ class std::vector < o2::tpc::Digit> + ;
+// pragma link C++ class std::vector < o2::tpc::PadFlags> + ;
 #pragma link C++ class std::vector < o2::tpc::ZeroSuppressedContainer8kb> + ;
 #pragma link C++ class o2::tpc::TrackTPC + ;
 #pragma link C++ class o2::tpc::LaserTrack + ;

--- a/DataFormats/Detectors/TPC/src/VectorPadflagsCustomStreamer.cxx
+++ b/DataFormats/Detectors/TPC/src/VectorPadflagsCustomStreamer.cxx
@@ -1,0 +1,56 @@
+// Copyright 2019-2025 CERN and copyright holders of ALICE O2.
+// See https://alice-o2.web.cern.ch/copyright for details of the copyright holders.
+// All rights not expressly granted are reserved.
+//
+// This software is distributed under the terms of the GNU General Public
+// License v3 (GPL Version 3), copied verbatim in the file "COPYING".
+//
+// In applying this license CERN does not waive the privileges and immunities
+// granted to it by virtue of its status as an Intergovernmental Organization
+// or submit itself to any jurisdiction.
+
+using std::vector;
+
+void VectorPadFlagsStreamer(TBuffer& R__b, void* objp)
+{
+  std::vector<o2::tpc::PadFlags>* obj = static_cast<std::vector<o2::tpc::PadFlags>*>(objp);
+  if (R__b.IsReading()) {
+    std::vector<int> R__stl;
+    R__stl.clear();
+    int R__n;
+    R__b >> R__n;
+    R__stl.reserve(R__n);
+    for (int R__i = 0; R__i < R__n; R__i++) {
+      Int_t readtemp;
+      R__b >> readtemp;
+      R__stl.push_back(readtemp);
+    }
+    auto data = reinterpret_cast<unsigned short*>(R__stl.data());
+    for (int i = 0; i < R__n; ++i) {
+      obj->push_back(static_cast<o2::tpc::PadFlags>(data[i]));
+    }
+  } else {
+    // We always save things with the old format.
+    R__b << (int)obj->size() / 2;
+    for (size_t i = 0; i < obj->size(); i++) {
+      R__b << (short)obj->at(i);
+    }
+  }
+}
+
+#define RootStreamerLocal(name, STREAMER)     \
+  namespace ROOT                              \
+  {                                           \
+                                              \
+  /** \cond HIDDEN_SYMBOLS */                 \
+  static auto _R__UNIQUE_(R__dummyStreamer) = \
+    []() { TClass::GetClass<name>()->SetStreamerFunc(STREAMER); return 0; }();                               \
+  /** \endcond */                             \
+  R__UseDummy(_R__UNIQUE_(R__dummyStreamer)); \
+  }
+
+// Let's not try to fix the old ROOT version, so that we can build
+// the new ROOT with the patched code in the CI.
+#if ROOT_VERSION_CODE >= ROOT_VERSION(6, 33, 00)
+RootStreamerLocal(vector<o2::tpc::PadFlags>, VectorPadFlagsStreamer);
+#endif

--- a/cmake/AddRootDictionary.cmake
+++ b/cmake/AddRootDictionary.cmake
@@ -51,7 +51,7 @@ function(add_root_dictionary target)
                         1
                         A
                         ""
-                        "LINKDEF"
+                        "LINKDEF;EXTRA_PATCH"
                         "HEADERS;BASENAME")
   if(A_UNPARSED_ARGUMENTS)
     message(
@@ -112,7 +112,7 @@ function(add_root_dictionary target)
   set(pcmBase ${dictionary}_rdict.pcm)
   set(pcmFile ${lib_output_dir}/${pcmBase})
   set(rootmapFile ${lib_output_dir}/lib${basename}.rootmap)
-  
+
   set(O2_TARGETPCMMAP_TARGET "${O2_TARGETPCMMAP_TARGET};${target}" CACHE INTERNAL "target/PCM map (target)")
   set(O2_TARGETPCMMAP_PCM "${O2_TARGETPCMMAP_PCM};${pcmFile}" CACHE INTERNAL "target/PCM map (pcm)")
 
@@ -132,6 +132,7 @@ function(add_root_dictionary target)
   set(includeDirs $<TARGET_PROPERTY:${target},INCLUDE_DIRECTORIES>)
   set(includeDirs $<REMOVE_DUPLICATES:${includeDirs}>)
 
+  list(LENGTH A_EXTRA_PATCH hasExtraPatch)
   # add a custom command to generate the dictionary using rootcling
   # cmake-format: off
   add_custom_command(
@@ -146,11 +147,13 @@ function(add_root_dictionary target)
       --include_dirs -I$<JOIN:${includeDirs},$<SEMICOLON>-I>
       $<$<BOOL:${prop}>:--compile_defs>
       $<$<BOOL:${prop}>:-D$<JOIN:${prop},$<SEMICOLON>-D>>
+      $<$<BOOL:${hasExtraPatch}>:--extra-patch>
+      $<$<BOOL:${hasExtraPatch}>:${CMAKE_CURRENT_LIST_DIR}/${A_EXTRA_PATCH}>
       --pcmdeps "$<REMOVE_DUPLICATES:${list_pcm_deps_${target}}>"
       --headers "${headers}"
     COMMAND
     ${CMAKE_COMMAND} -E copy_if_different ${CMAKE_CURRENT_BINARY_DIR}/${pcmBase} ${pcmFile}
-    DEPENDS ${headers} "$<REMOVE_DUPLICATES:${list_pcm_deps_${target}}>")
+    DEPENDS ${headers} "$<REMOVE_DUPLICATES:${list_pcm_deps_${target}}>" ${A_EXTRA_PATCH})
   # cmake-format: on
 
   # add dictionary source to the target sources

--- a/cmake/O2TargetRootDictionary.cmake
+++ b/cmake/O2TargetRootDictionary.cmake
@@ -55,13 +55,8 @@ function(o2_target_root_dictionary baseTargetName)
                         1
                         A
                         ""
-                        "LINKDEF"
+                        "LINKDEF;EXTRA_PATCH"
                         "HEADERS")
-
-  if(A_UNPARSED_ARGUMENTS)
-    message(
-      FATAL_ERROR "Unexpected unparsed arguments: ${A_UNPARSED_ARGUMENTS}")
-  endif()
 
   if(${ARGC} LESS 1)
     message(
@@ -96,6 +91,13 @@ function(o2_target_root_dictionary baseTargetName)
 
   # now that we have the O2 specific stuff computed, delegate the actual work to
   # the add_root_dictionary function
-  add_root_dictionary(${target} HEADERS ${A_HEADERS} LINKDEF ${A_LINKDEF})
+if(NOT A_EXTRA_PATCH)
+    add_root_dictionary(${target} HEADERS ${A_HEADERS} LINKDEF ${A_LINKDEF})
+else()
+    add_root_dictionary(${target}
+                        EXTRA_PATCH ${A_EXTRA_PATCH}
+                        HEADERS ${A_HEADERS}
+                        LINKDEF ${A_LINKDEF})
+endif()
 
 endfunction()

--- a/cmake/rootcling_wrapper.sh.in
+++ b/cmake/rootcling_wrapper.sh.in
@@ -41,6 +41,10 @@ while [[ $# -gt 0 ]]; do
     PCMDEPS="$2"
     shift 2
     ;;
+  --extra-patch)
+    EXTRA_PATCH="$2"
+    shift 2
+    ;;
   *)
     if [[ -z "$1" ]]; then
       shift
@@ -82,6 +86,18 @@ esac
 
 LOGFILE=${DICTIONARY_FILE}.log
 
+echo @CMAKE_COMMAND@ -E env "LD_LIBRARY_PATH=$libpath" @ROOT_rootcling_CMD@ \
+  -f $DICTIONARY_FILE \
+  -inlineInputHeader \
+  -noGlobalUsingStd \
+  -rmf ${ROOTMAP_FILE} \
+  -rml ${ROOTMAP_LIBRARY_NAME} \
+  ${INCLUDE_DIRS//;/ } \
+  ${COMPILE_DEFINITIONS//;/ } \
+  ${PCMDEPS:+-m }${PCMDEPS//;/ -m } \
+  ${HEADERS//;/ } \
+  > ${LOGFILE} 2>&1 || ROOTCLINGRETVAL=$?
+
 @CMAKE_COMMAND@ -E env "LD_LIBRARY_PATH=$libpath" @ROOT_rootcling_CMD@ \
   -f $DICTIONARY_FILE \
   -inlineInputHeader \
@@ -93,6 +109,13 @@ LOGFILE=${DICTIONARY_FILE}.log
   ${PCMDEPS:+-m }${PCMDEPS//;/ -m } \
   ${HEADERS//;/ } \
   > ${LOGFILE} 2>&1 || ROOTCLINGRETVAL=$?
+
+# Add the extra patch file at the end of the generated dictionary.
+# This is needed to inject custom streamers (e.g. for std::vector<PadFlags>)
+# to our dictionary.
+if [ ! X"${EXTRA_PATCH}" = X ]; then
+  cat $EXTRA_PATCH >> ${DICTIONARY_FILE}
+fi
 
 if [[ ${ROOTCLINGRETVAL:-0} != "0" ]]; then
   cat ${LOGFILE} >&2

